// @libria/gridstack/esm5/lib/components/gridstack-item.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/components/gridstack-item.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, Input, Renderer2, ViewEncapsulation, } from '@angular/core';
/** @type {?} */
var _sequence = 0;
var GridstackItemComponent = /** @class */ (function () {
    function GridstackItemComponent(elem, _renderer) {
        this.elem = elem;
        this._renderer = _renderer;
        this.generatedId = (_sequence++).toString();
    }
    /**
     * @return {?}
     */
    GridstackItemComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        this._renderer.addClass(this.elem.nativeElement, 'grid-stack-item');
        this._setAttributeIfNotUndefined('data-gs-x', this.x);
        this._setAttributeIfNotUndefined('data-gs-y', this.y);
        this._setAttributeIfNotUndefined('data-gs-width', this.width);
        this._setAttributeIfNotUndefined('data-gs-height', this.height);
        this._setAttributeIfNotUndefined('data-gs-id', this.id);
        this._setAttributeIfNotUndefined('data-gs-max-width', this.maxWidth);
        this._setAttributeIfNotUndefined('data-gs-max-height', this.maxHeight);
        this._setAttributeIfNotUndefined('data-gs-min-width', this.minWidth);
        this._setAttributeIfNotUndefined('data-gs-min-height', this.minHeight);
        this._setAttributeIfTrue('data-gs-no-resize', this.noResize);
        this._setAttributeIfTrue('data-gs-no-move', this.noMove);
        this._setAttributeIfTrue('data-gs-auto-position', this.autoPosition);
        this._setAttributeIfTrue('data-gs-locked', this.locked);
    };
    /**
     * @private
     * @param {?} attrName
     * @param {?} val
     * @return {?}
     */
    GridstackItemComponent.prototype._setAttributeIfNotUndefined = /**
     * @private
     * @param {?} attrName
     * @param {?} val
     * @return {?}
     */
    function (attrName, val) {
        if (!_.isNull(val) && !_.isUndefined(val)) {
            this._renderer.setAttribute(this.elem.nativeElement, attrName, val.toString());
        }
    };
    /**
     * @private
     * @param {?} attrName
     * @param {?} val
     * @return {?}
     */
    GridstackItemComponent.prototype._setAttributeIfTrue = /**
     * @private
     * @param {?} attrName
     * @param {?} val
     * @return {?}
     */
    function (attrName, val) {
        if (val === true || val === 'true') {
            this._renderer.setAttribute(this.elem.nativeElement, attrName, 'true');
        }
    };
    GridstackItemComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line:component-selector
                    selector: 'div[lb-gridstack-item]',
                    template: "<div class=\"grid-stack-item-content\">\r\n    <ng-content></ng-content>\r\n</div>",
                    encapsulation: ViewEncapsulation.None,
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    GridstackItemComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    GridstackItemComponent.propDecorators = {
        x: [{ type: Input }],
        y: [{ type: Input }],
        width: [{ type: Input }],
        height: [{ type: Input }],
        id: [{ type: Input }],
        maxWidth: [{ type: Input, args: ['max-width',] }],
        maxHeight: [{ type: Input, args: ['max-height',] }],
        minWidth: [{ type: Input, args: ['min-width',] }],
        minHeight: [{ type: Input, args: ['min-height',] }],
        noResize: [{ type: Input, args: ['no-resize',] }],
        noMove: [{ type: Input, args: ['no-move',] }],
        autoPosition: [{ type: Input, args: ['auto-position',] }],
        locked: [{ type: Input }]
    };
    return GridstackItemComponent;
}());
export { GridstackItemComponent };
if (false) {
    /** @type {?} */
    GridstackItemComponent.prototype.x;
    /** @type {?} */
    GridstackItemComponent.prototype.y;
    /** @type {?} */
    GridstackItemComponent.prototype.width;
    /** @type {?} */
    GridstackItemComponent.prototype.height;
    /** @type {?} */
    GridstackItemComponent.prototype.id;
    /** @type {?} */
    GridstackItemComponent.prototype.maxWidth;
    /** @type {?} */
    GridstackItemComponent.prototype.maxHeight;
    /** @type {?} */
    GridstackItemComponent.prototype.minWidth;
    /** @type {?} */
    GridstackItemComponent.prototype.minHeight;
    /** @type {?} */
    GridstackItemComponent.prototype.noResize;
    /** @type {?} */
    GridstackItemComponent.prototype.noMove;
    /** @type {?} */
    GridstackItemComponent.prototype.autoPosition;
    /** @type {?} */
    GridstackItemComponent.prototype.locked;
    /** @type {?} */
    GridstackItemComponent.prototype.generatedId;
    /** @type {?} */
    GridstackItemComponent.prototype.elem;
    /**
     * @type {?}
     * @private
     */
    GridstackItemComponent.prototype._renderer;
}
