// @libria/gridstack/esm5/lib/services/gridstack.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/services/gridstack.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
var GridstackService = /** @class */ (function () {
    function GridstackService() {
        this._gridItems = [];
        this._grids = [];
    }
    /**
     * @param {?} grid
     * @return {?}
     */
    GridstackService.prototype.addGrid = /**
     * @param {?} grid
     * @return {?}
     */
    function (grid) {
        this._grids.push(grid);
    };
    /**
     * @param {?} grid
     * @return {?}
     */
    GridstackService.prototype.removeGrid = /**
     * @param {?} grid
     * @return {?}
     */
    function (grid) {
        this._grids.splice(this._grids.findIndex((/**
         * @param {?} g
         * @return {?}
         */
        function (g) { return g.generatedId == grid.generatedId; })), 1);
        this._gridItems = this._gridItems.filter((/**
         * @param {?} gi
         * @return {?}
         */
        function (gi) { return gi.gridId != grid.generatedId; }));
    };
    /**
     * @param {?} gridId
     * @param {?} gridItemId
     * @return {?}
     */
    GridstackService.prototype.attachGridItem = /**
     * @param {?} gridId
     * @param {?} gridItemId
     * @return {?}
     */
    function (gridId, gridItemId) {
        this._gridItems.find((/**
         * @param {?} gi
         * @return {?}
         */
        function (gi) { return gi.item.generatedId == gridItemId; })).gridId = gridId;
    };
    /**
     * @param {?} gridItemId
     * @return {?}
     */
    GridstackService.prototype.detachGridItemIfExists = /**
     * @param {?} gridItemId
     * @return {?}
     */
    function (gridItemId) {
        /** @type {?} */
        var gridItem = this._gridItems.find((/**
         * @param {?} gi
         * @return {?}
         */
        function (gi) { return gi.item.generatedId == gridItemId; }));
        if (!!gridItem) {
            gridItem.gridId = null;
        }
    };
    /**
     * @param {?} gridId
     * @return {?}
     */
    GridstackService.prototype.getGridItems = /**
     * @param {?} gridId
     * @return {?}
     */
    function (gridId) {
        return this._gridItems.filter((/**
         * @param {?} gi
         * @return {?}
         */
        function (gi) { return gi.gridId == gridId; })).map((/**
         * @param {?} g
         * @return {?}
         */
        function (g) { return g.item; }));
    };
    /**
     * @return {?}
     */
    GridstackService.prototype.getOrphanGridItems = /**
     * @return {?}
     */
    function () {
        return this._gridItems.filter((/**
         * @param {?} gi
         * @return {?}
         */
        function (gi) { return gi.gridId == null; })).map((/**
         * @param {?} g
         * @return {?}
         */
        function (g) { return g.item; }));
    };
    /**
     * @param {?} gridItem
     * @return {?}
     */
    GridstackService.prototype.addOrphanGridItem = /**
     * @param {?} gridItem
     * @return {?}
     */
    function (gridItem) {
        this._gridItems.push({
            gridId: null,
            item: gridItem
        });
    };
    /**
     * @param {?} gridItemId
     * @return {?}
     */
    GridstackService.prototype.removeGridItem = /**
     * @param {?} gridItemId
     * @return {?}
     */
    function (gridItemId) {
        this._gridItems.splice(this._gridItems.findIndex((/**
         * @param {?} gi
         * @return {?}
         */
        function (gi) { return gi.item.generatedId == gridItemId; })), 1);
    };
    GridstackService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    GridstackService.ctorParameters = function () { return []; };
    return GridstackService;
}());
export { GridstackService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    GridstackService.prototype._gridItems;
    /**
     * @type {?}
     * @private
     */
    GridstackService.prototype._grids;
}
