// @libria/gridstack/esm5/lib/gridstack.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/gridstack.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { GridstackComponent } from './components/gridstack.component';
import { GridstackItemComponent } from './components/gridstack-item.component';
import { GridstackService } from './services/gridstack.service';
var GridstackModule = /** @class */ (function () {
    function GridstackModule() {
    }
    /**
     * @return {?}
     */
    GridstackModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: GridstackModule,
            providers: [
                GridstackService
            ]
        };
    };
    GridstackModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule
                    ],
                    declarations: [
                        GridstackComponent,
                        GridstackItemComponent
                    ],
                    exports: [
                        GridstackComponent,
                        GridstackItemComponent
                    ]
                },] }
    ];
    return GridstackModule;
}());
export { GridstackModule };
