// @libria/gridstack/esm2015/lib/services/gridstack.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/services/gridstack.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
export class GridstackService {
    constructor() {
        this._gridItems = [];
        this._grids = [];
    }
    /**
     * @param {?} grid
     * @return {?}
     */
    addGrid(grid) {
        this._grids.push(grid);
    }
    /**
     * @param {?} grid
     * @return {?}
     */
    removeGrid(grid) {
        this._grids.splice(this._grids.findIndex((/**
         * @param {?} g
         * @return {?}
         */
        g => g.generatedId == grid.generatedId)), 1);
        this._gridItems = this._gridItems.filter((/**
         * @param {?} gi
         * @return {?}
         */
        gi => gi.gridId != grid.generatedId));
    }
    /**
     * @param {?} gridId
     * @param {?} gridItemId
     * @return {?}
     */
    attachGridItem(gridId, gridItemId) {
        this._gridItems.find((/**
         * @param {?} gi
         * @return {?}
         */
        gi => gi.item.generatedId == gridItemId)).gridId = gridId;
    }
    /**
     * @param {?} gridItemId
     * @return {?}
     */
    detachGridItemIfExists(gridItemId) {
        /** @type {?} */
        const gridItem = this._gridItems.find((/**
         * @param {?} gi
         * @return {?}
         */
        gi => gi.item.generatedId == gridItemId));
        if (!!gridItem) {
            gridItem.gridId = null;
        }
    }
    /**
     * @param {?} gridId
     * @return {?}
     */
    getGridItems(gridId) {
        return this._gridItems.filter((/**
         * @param {?} gi
         * @return {?}
         */
        gi => gi.gridId == gridId)).map((/**
         * @param {?} g
         * @return {?}
         */
        g => g.item));
    }
    /**
     * @return {?}
     */
    getOrphanGridItems() {
        return this._gridItems.filter((/**
         * @param {?} gi
         * @return {?}
         */
        gi => gi.gridId == null)).map((/**
         * @param {?} g
         * @return {?}
         */
        g => g.item));
    }
    /**
     * @param {?} gridItem
     * @return {?}
     */
    addOrphanGridItem(gridItem) {
        this._gridItems.push({
            gridId: null,
            item: gridItem
        });
    }
    /**
     * @param {?} gridItemId
     * @return {?}
     */
    removeGridItem(gridItemId) {
        this._gridItems.splice(this._gridItems.findIndex((/**
         * @param {?} gi
         * @return {?}
         */
        gi => gi.item.generatedId == gridItemId)), 1);
    }
}
GridstackService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
GridstackService.ctorParameters = () => [];
if (false) {
    /**
     * @type {?}
     * @private
     */
    GridstackService.prototype._gridItems;
    /**
     * @type {?}
     * @private
     */
    GridstackService.prototype._grids;
}
